Fix #25099 set na_rep values before converting to string to prevent data truncation

kinow opened the pull request:
Hi,

The code in this pull request fixes #25099, and all tests pass. However, I am not sure if this is the best (or even an elegant) solution.

The issue happened because the float `NaN` is converted to `str` (i.e. "nan"), which gets stored as a `<U3`. When the user-provided `na_rep` is used, it is first truncated down to the data type length (i.e. "myn").

I tried moving the `na_rep` to before it is converted to string, but other tests fail. I could not find other code using a similar approach, of concatenating `<U` with the desired length. So I assume it's an easy hack, but not a proper solution?

Any suggestions? I also tried setting the `na_rep` value only in the `if` branch that calls `values.astype(str)`, but there are still test failures.

Cheers
Bruno

- [x] closes #25099
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

kinow committed:
Fix #25099 set na_rep values before converting to string to prevent data truncation

jreback commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -561,3 +561,19 @@ def test_to_csv_compression(self, compression_only,
+                         'mynull',
+                         'mynull']
+        expected_ascii = tm.convert_rows_list_to_csv_str(expected_rows)
+
```
just write to a StringIO, no need to use stdout here.

use result= and expected=

kinow replied:
Sure thing! Updating the branch in a sec. Thanks @jreback !

kinow replied:
Done!

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -739,7 +739,10 @@ def to_native_types(self, slicer=None, na_rep='nan', quoting=None,
 
         if not self.is_object and not quoting:
-            values = values.astype(str)
+            if na_rep and isinstance(na_rep, str):
```
this is not going to work, see how we do this in ``pandas/io/pytables`` in ``_convert_string_array``. we will want to adopt that approach.

kinow replied:
Tried reusing `_convert_string_array` first to see if that would work, but it returns `S{...}`, which causes tests to fail (e.g. expected `'mynull'`, instead got `b'mynull'`).

Then tried to re-use parts of the function. More specifically the part that calculates the text length (`bytes` or `string`). I created a function to be used in both `_convert_string_array` and by other callers.

Alas, I couldn't get it to work as a `cdef` (I believe Python Python cannot call it?), nor as `cpdef` (which was funny, but some tests were failing with being aborted with `SIGABRT` :confused: ). In the end, I only managed to get it to work by having a simple function, without type hinting.

@jreback probably not what you had in mind. Care to take a look and suggest an alternative whenever you have time, please? Pushed in a separate commit.

Cheers
Bruno

pep8speaks commented:
Hello @kinow! Thanks for updating the PR.



 - In the file [`pandas/tests/io/formats/test_to_csv.py`](https://github.com/pandas-dev/pandas/blob/388cd596ae626aeb58a7b7030c061de770eaccb2/pandas/tests/io/formats/test_to_csv.py), following are the PEP8 issues :

> [Line 582:1](https://github.com/pandas-dev/pandas/blob/388cd596ae626aeb58a7b7030c061de770eaccb2/pandas/tests/io/formats/test_to_csv.py#L582): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

jreback commented on pandas/_libs/writers.pyx:
```diff
@@ -128,19 +128,28 @@ def max_len_string_array(pandas_string[:] arr) -> Py_ssize_t:
     return m
 
 
+def max_len_string(val):
```
this needs to be ``cdef inline Py_ssize_t max_len_string(object val)``

kinow replied:
Done! Yesterday I tried

`cdef Py_ssize_t max_len_string(object val)`
`cdef Py_ssize_t max_len_string(pandas_string val)`
`cpdef max_len_string(object val) -> Py_ssize_t` and 
`cpdef max_len_string(object val) -> Py_ssize_t`

And ended up giving up and reverting to the simpler signature above. I think `pytables` tests were failing in my IDE, but I didn't try on Travis CI. Pushed this change now to see if that passes the tests.

Thanks a ton @jreback ! Still need to run the asv's you mentioned above.

kinow replied:
Same error I had @jreback : https://travis-ci.org/kinow/pandas/jobs/488681016

```
AttributeError: module 'pandas._libs.writers' has no attribute 'max_len_string'
```

Is it because the `cdef` function cannot be called from Python? ([this](https://notes-on-cython.readthedocs.io/en/latest/function_declarations.html) is what I used yesterday trying to understand why I had this error). Then after that I tried `cpdef` and changed where the return type was defined, but that resulted in another error..

jreback replied:
correct, you can mark it cpdef to get both python and cython functionaility

kinow replied:
You nailed it @jreback ! It worked! 👏 🎉

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -739,7 +739,11 @@ def to_native_types(self, slicer=None, na_rep='nan', quoting=None,
 
         if not self.is_object and not quoting:
-            values = values.astype(str)
+            if na_rep is not None and isinstance(na_rep, (str, bytes)):
```
na_rep can *only* be a string; I think you are safe to only have the fixed string formatting

jreback commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -561,3 +563,19 @@ def test_to_csv_compression(self, compression_only,
+        expected_ascii = tm.convert_rows_list_to_csv_str(expected_rows)
+
+        df.to_csv(sys.stdout, index=False, na_rep='mynull', encoding='ascii')
+        captured = capsys.readouterr()
```
you need to use StringIO here

kinow replied:
Oh, I think I missed it when squashing commits and merging, even though I left the line `from pandas.compat import StringIO` at the top. Thanks for the spot. Let me fix this one now before I forget it again.

jreback reviewed (changes requested):
pls run the asv's for csv and pytables as well and report if any regressions

kinow commented:
@jreback just finished running asv with the following command

```bash
asv continuous -f 1.1 upstream/master HEAD -b ^io.csv
```

Complete output in this [gist](https://gist.github.com/kinow/9e3760ddfa036b82df0f587323e17ee6), but near the bottom of its output, it says

>+      9.92±0.1ms       11.7±0.6ms     1.18  io.csv.ReadCSVFloatPrecision.time_read_csv_python_engine(',', >'_', None)
>+         119±1ms          135±4ms     1.14  io.csv.ReadCSVCategorical.time_convert_direct
>+         171±5ms          190±4ms     1.11  io.csv.ReadCSVCategorical.time_convert_post
>
>SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.

Which doesn't look good I think... do you reckon I should try to find another alternative, perhaps duplicating code and leaving the previous method unaltered? 

Thanks
Bruno

jreback commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -561,3 +563,19 @@ def test_to_csv_compression(self, compression_only,
                                  compression=read_compression)
             tm.assert_frame_equal(result, df)
+
+    def test_to_csv_na_rep_long_string(self, capsys):
```
you can remove the capsys

kinow replied:
Good spot!

jreback commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -561,3 +563,19 @@ def test_to_csv_compression(self, compression_only,
+    def test_to_csv_na_rep_long_string(self, capsys):
+        # see gh-25099
+        df = pd.DataFrame({"c": [float('nan')] * 3})
+        df = df.astype("Int64")
```
can you parametrize this function on ``float`` and ``Ing64``

kinow replied:
Good idea. Done!

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -739,7 +739,8 @@ def to_native_types(self, slicer=None, na_rep='nan', quoting=None,
 
         if not self.is_object and not quoting:
-            values = values.astype(str)
+            itemsize = writers.max_len_string(na_rep)
```
maybe a better way is to do the ``values[mask] = na_rep`` before doing what we were doing originally which is the ``.astype(str)`` (which I think is pretty fast).

you prob need to least the ``values[mask] = na_rep`` in the else block though.

kinow replied:
Oh, I think it could work. 

>you prob need to least the `values[mask] = na_rep` in the else block though.

I moved the `values[mask] = na_rep` before the `if/else`, then left the `values.astype(str)` and `np.array(values, dtype='object')` to take care of the typing. It appears to have worked, at least from the couple of unit test classes I had open in the IDE for testing.

Running `test_fast.sh` now, and if that works, will revert the cython change, push to see if Travis is happy, and in parallel run the `asv` test again and report back its results.

jreback replied:
yes this is somewhat perf sensitive code, so IIRC that's why this has this kind of ordereing, but will see.

kinow replied:
@jreback 

The change I did was

```diff
diff --git a/pandas/core/internals/blocks.py b/pandas/core/internals/blocks.py
index 1bf122558..542925e88 100644
--- a/pandas/core/internals/blocks.py
+++ b/pandas/core/internals/blocks.py
@@ -738,13 +738,12 @@ class Block(PandasObject):
             values = values[:, slicer]
         mask = isna(values)
 
+        values[mask] = na_rep
         if not self.is_object and not quoting:
-            itemsize = writers.max_len_string(na_rep)
-            values = values.astype("<U{size}".format(size=itemsize))
+            values = values.astype(str)
         else:
             values = np.array(values, dtype='object')
 
-        values[mask] = na_rep
         return values
 
     # block actions ####
```

Resulted in

```
============================================= 95 failed, 45225 passed, 1185 skipped, 1000 xfailed, 22 xpassed, 121 warnings, 23 error in 440.68 seconds =============================================
```

Then tried this one

```diff
diff --git a/pandas/core/internals/blocks.py b/pandas/core/internals/blocks.py
index 1bf122558..0aa744071 100644
--- a/pandas/core/internals/blocks.py
+++ b/pandas/core/internals/blocks.py
@@ -739,12 +739,12 @@ class Block(PandasObject):
         mask = isna(values)
 
         if not self.is_object and not quoting:
-            itemsize = writers.max_len_string(na_rep)
-            values = values.astype("<U{size}".format(size=itemsize))
+            values[mask] = na_rep
+            values = values.astype(str)
         else:
             values = np.array(values, dtype='object')
+            values[mask] = na_rep
 
-        values[mask] = na_rep
         return values
 
     # block actions ####
```

Resulted in 

```
============================================= 93 failed, 45227 passed, 1185 skipped, 1000 xfailed, 22 xpassed, 120 warnings, 23 error in 456.06 seconds =============================================

```

I've got a few tests (19) that fail in my environment even when Travis CI build passes, so am pushing the second version to git. Is it similar to what you had in mind @jreback ?

kinow replied:
Looks like Travis had some tests failing after my last commit...

jreback commented:
can you merge master

kinow commented:
Branch rebased onto `master`

jreback reviewed (changes requested):
can you add a whatsnew note in bug fixes, I/O section. ping on green.

kinow committed:
Add whatsnew

kinow commented:
@jreback added a whatsnew entry as requested - happy to amend if text is incorrect or too simple.

I resolved the outdated discussions, except for one that is still pertinent. The initial approach was to calculate the maximum length when it was a string, and then just create a type of that length. That way the masked values were not truncated.

The pending discussion is around that approach, where we tried instead to change the order of where the mask gets applied. But several other tests failed after that (see my last comment that Travis failed). In one test that I found, it had an empty string, and for some reason with the different order, it was trying to call `int('')` and returning `ValueError: invalid literal for int() with base 10: ''`.

So I reverted that change, and Travis stopped complaining. Also ran `asv`:

```
$ asv continuous -f 1.1 upstream/master HEAD -b ^io.csv
· No information stored about machine 'ranma'. I know about nothing.
  

I will now ask you some questions about this machine to identify it in the benchmarks.

1. machine:  A unique name to identify this machine in the results.
   May be anything, as long as it is unique across all the machines
   used to benchmark this project.  NOTE: If changed from the default,
   it will no longer match the hostname of this machine, and you may
   need to explicitly use the --machine argument to asv.
machine [ranma]: 
2. os:  The OS type and version of this machine.  For example,
   'Macintosh OS-X 10.8'.
os [Linux 4.15.0-46-generic]: 
3. arch:  The generic CPU architecture of this machine.  For example,
   'i386' or 'x86_64'.
arch [x86_64]: 
4. cpu:  A specific description of the CPU of this machine, including
   its speed and class.  For example, 'Intel(R) Core(TM) i5-2520M CPU
   @ 2.50GHz (4 cores)'.
cpu [Intel(R) Core(TM) i7-3720QM CPU @ 2.60GHz]: 
5. ram:  The amount of physical RAM on this machine.  For example,
   '4GB'.
ram [16119292]: 
· Creating environments......................................................................................................................................................................................................
· Discovering benchmarks
·· Uninstalling from conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
·· Building 6e0dd9cf <fix-issue-gh25099-csv-narep> for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.........................................................
·· Installing 6e0dd9cf <fix-issue-gh25099-csv-narep> into conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
· Running 32 total benchmarks (2 commits * 1 environments * 16 benchmarks)
[  0.00%] · For pandas commit 46639512 <master> (round 1/2):
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt............................................................
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  1.56%] ··· Running (io.csv.ReadCSVCategorical.time_convert_direct--)...............
[ 25.00%] · For pandas commit 6e0dd9cf <fix-issue-gh25099-csv-narep> (round 1/2):
[ 25.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 25.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 26.56%] ··· Running (io.csv.ReadCSVCategorical.time_convert_direct--)...............
[ 50.00%] · For pandas commit 6e0dd9cf <fix-issue-gh25099-csv-narep> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 51.56%] ··· io.csv.ReadCSVCategorical.time_convert_direct                                                                                                          56.9±0.6ms
[ 53.12%] ··· io.csv.ReadCSVCategorical.time_convert_post                                                                                                            82.8±0.5ms
[ 54.69%] ··· io.csv.ReadCSVComment.time_comment                                                                                                                     38.6±0.6ms
[ 56.25%] ··· io.csv.ReadCSVDInferDatetimeFormat.time_read_csv                                                                                                               ok
[ 56.25%] ··· ======================= ============= ============= =============
              --                                        format                 
              ----------------------- -----------------------------------------
               infer_datetime_format      custom       iso8601         ymd     
              ======================= ============= ============= =============
                        True           8.42±0.02ms   2.87±0.03ms   2.89±0.01ms 
                       False             138±1ms     2.41±0.01ms   2.27±0.01ms 
              ======================= ============= ============= =============

[ 57.81%] ··· io.csv.ReadCSVFloatPrecision.time_read_csv                                                                                                                     ok
[ 57.81%] ··· ===== ============= ============= ================ ============= ============= ================
              --                                    decimal / float_precision                                
              ----- -----------------------------------------------------------------------------------------
               sep     . / None      . / high    . / round_trip     _ / None      _ / high    _ / round_trip 
              ===== ============= ============= ================ ============= ============= ================
                ,    2.31±0.02ms   2.21±0.04ms    2.95±0.03ms     2.56±0.03ms   2.56±0.02ms    2.56±0.03ms   
                ;    2.33±0.02ms   2.24±0.05ms    2.96±0.04ms     2.53±0.03ms   2.49±0.04ms    2.53±0.03ms   
              ===== ============= ============= ================ ============= ============= ================

[ 59.38%] ··· io.csv.ReadCSVFloatPrecision.time_read_csv_python_engine                                                                                                       ok
[ 59.38%] ··· ===== ============ ============ ================ ============= ============= ================
              --                                   decimal / float_precision                               
              ----- ---------------------------------------------------------------------------------------
               sep    . / None     . / high    . / round_trip     _ / None      _ / high    _ / round_trip 
              ===== ============ ============ ================ ============= ============= ================
                ,    4.66±0.2ms   4.63±0.1ms     4.70±0.1ms     3.90±0.05ms   3.90±0.03ms    3.89±0.05ms   
                ;    4.64±0.2ms   4.62±0.1ms     4.62±0.2ms     3.88±0.02ms   3.89±0.05ms    3.90±0.04ms   
              ===== ============ ============ ================ ============= ============= ================

[ 60.94%] ··· io.csv.ReadCSVMemoryGrowth.mem_parser_chunks                                                                                                                    0
[ 62.50%] ··· io.csv.ReadCSVParseDates.time_baseline                                                                                                                2.71±0.05ms
[ 64.06%] ··· io.csv.ReadCSVParseDates.time_multiple_date                                                                                                           3.13±0.01ms
[ 65.62%] ··· io.csv.ReadCSVSkipRows.time_skipprows                                                                                                                          ok
[ 65.62%] ··· ========== ============
               skiprows              
              ---------- ------------
                 None     23.0±0.2ms 
                10000     15.1±0.3ms 
              ========== ============

[ 67.19%] ··· io.csv.ReadCSVThousands.time_thousands                                                                                                                         ok
[ 67.19%] ··· ===== ============ ============
              --            thousands        
              ----- -------------------------
               sep      None          ,      
              ===== ============ ============
                ,    18.6±0.2ms   19.5±0.5ms 
                |    18.6±0.3ms   20.7±0.2ms 
              ===== ============ ============

[ 68.75%] ··· io.csv.ReadUint64Integers.time_read_uint64                                                                                                             4.79±0.1ms
[ 70.31%] ··· io.csv.ReadUint64Integers.time_read_uint64_na_values                                                                                                  7.47±0.07ms
[ 71.88%] ··· io.csv.ReadUint64Integers.time_read_uint64_neg_values                                                                                                 7.32±0.09ms
[ 73.44%] ··· io.csv.ToCSV.time_frame                                                                                                                                        ok
[ 73.44%] ··· ======= ============
                kind              
              ------- ------------
                wide   133±0.8ms  
                long    237±1ms   
               mixed   29.8±0.6ms 
              ======= ============

[ 75.00%] ··· io.csv.ToCSVDatetime.time_frame_date_formatting                                                                                                        25.5±0.2ms
[ 75.00%] · For pandas commit 46639512 <master> (round 2/2):
[ 75.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 75.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 76.56%] ··· io.csv.ReadCSVCategorical.time_convert_direct                                                                                                          53.8±0.3ms
[ 78.12%] ··· io.csv.ReadCSVCategorical.time_convert_post                                                                                                            80.2±0.5ms
[ 79.69%] ··· io.csv.ReadCSVComment.time_comment                                                                                                                     39.6±0.7ms
[ 81.25%] ··· io.csv.ReadCSVDInferDatetimeFormat.time_read_csv                                                                                                               ok
[ 81.25%] ··· ======================= ============= ============= =============
              --                                        format                 
              ----------------------- -----------------------------------------
               infer_datetime_format      custom       iso8601         ymd     
              ======================= ============= ============= =============
                        True           8.48±0.06ms   2.89±0.02ms   2.87±0.03ms 
                       False            140±0.3ms    2.42±0.01ms   2.31±0.01ms 
              ======================= ============= ============= =============

[ 82.81%] ··· io.csv.ReadCSVFloatPrecision.time_read_csv                                                                                                                     ok
[ 82.81%] ··· ===== ============= ============= ================ ============= ============= ================
              --                                    decimal / float_precision                                
              ----- -----------------------------------------------------------------------------------------
               sep     . / None      . / high    . / round_trip     _ / None      _ / high    _ / round_trip 
              ===== ============= ============= ================ ============= ============= ================
                ,    2.31±0.04ms   2.22±0.03ms    3.00±0.04ms     2.58±0.04ms   2.58±0.04ms    2.63±0.05ms   
                ;    2.32±0.05ms   2.21±0.04ms    3.00±0.04ms     2.62±0.04ms   2.63±0.03ms    2.59±0.04ms   
              ===== ============= ============= ================ ============= ============= ================

[ 84.38%] ··· io.csv.ReadCSVFloatPrecision.time_read_csv_python_engine                                                                                                       ok
[ 84.38%] ··· ===== ============ ============ ================ ============ ============ ================
              --                                  decimal / float_precision                              
              ----- -------------------------------------------------------------------------------------
               sep    . / None     . / high    . / round_trip    _ / None     _ / high    _ / round_trip 
              ===== ============ ============ ================ ============ ============ ================
                ,    5.10±0.6ms   5.07±0.6ms     5.06±0.5ms     4.43±0.5ms   4.36±0.5ms     4.48±0.6ms   
                ;    5.07±0.6ms   5.05±0.5ms     5.09±0.5ms     4.41±0.5ms   4.36±0.5ms     4.41±0.6ms   
              ===== ============ ============ ================ ============ ============ ================

[ 85.94%] ··· io.csv.ReadCSVMemoryGrowth.mem_parser_chunks                                                                                                                    0
[ 87.50%] ··· io.csv.ReadCSVParseDates.time_baseline                                                                                                                2.72±0.05ms
[ 89.06%] ··· io.csv.ReadCSVParseDates.time_multiple_date                                                                                                           3.13±0.02ms
[ 90.62%] ··· io.csv.ReadCSVSkipRows.time_skipprows                                                                                                                          ok
[ 90.62%] ··· ========== ============
               skiprows              
              ---------- ------------
                 None     22.9±0.3ms 
                10000     15.0±0.1ms 
              ========== ============

[ 92.19%] ··· io.csv.ReadCSVThousands.time_thousands                                                                                                                         ok
[ 92.19%] ··· ===== ============ ============
              --            thousands        
              ----- -------------------------
               sep      None          ,      
              ===== ============ ============
                ,    18.9±0.2ms   19.6±0.2ms 
                |    19.1±0.2ms   20.6±0.1ms 
              ===== ============ ============

[ 93.75%] ··· io.csv.ReadUint64Integers.time_read_uint64                                                                                                            4.75±0.06ms
[ 95.31%] ··· io.csv.ReadUint64Integers.time_read_uint64_na_values                                                                                                  7.38±0.08ms
[ 96.88%] ··· io.csv.ReadUint64Integers.time_read_uint64_neg_values                                                                                                  7.32±0.1ms
[ 98.44%] ··· io.csv.ToCSV.time_frame                                                                                                                                        ok
[ 98.44%] ··· ======= ============
                kind              
              ------- ------------
                wide   130±0.4ms  
                long    234±5ms   
               mixed   28.9±0.4ms 
              ======= ============

[100.00%] ··· io.csv.ToCSVDatetime.time_frame_date_formatting                                                                                                        25.7±0.5ms

BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

However, I am still not too confident if that is a valid approach, as I did not find other parts of the code in Pandas where types had a dynamic size :+1:

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -271,6 +271,7 @@ I/O
 - Bug in :func:`json_normalize` for ``errors='ignore'`` where missing values in the input data, were filled in resulting ``DataFrame`` with the string "nan" instead of ``numpy.nan`` (:issue:`25468`)
 - :meth:`DataFrame.to_html` now raises ``TypeError`` when using an invalid type for the ``classes`` parameter instead of ``AsseertionError`` (:issue:`25608`)
 - Bug in :meth:`DataFrame.to_string` and :meth:`DataFrame.to_latex` that would lead to incorrect output when the ``header`` keyword is used (:issue:`16718`)
+- Bug in data being truncated with ``na_pep`` on the length of the first value found. (:issue:`25099`)
```
this is mispelled, and I don't know which function this refers.

jreback replied:
use ``:meth:DataFrame.to_csv``

jreback reviewed (changes requested):
ping on green.

WillAyd commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -271,6 +271,7 @@ I/O
 - Bug in :func:`json_normalize` for ``errors='ignore'`` where missing values in the input data, were filled in resulting ``DataFrame`` with the string "nan" instead of ``numpy.nan`` (:issue:`25468`)
 - :meth:`DataFrame.to_html` now raises ``TypeError`` when using an invalid type for the ``classes`` parameter instead of ``AsseertionError`` (:issue:`25608`)
 - Bug in :meth:`DataFrame.to_string` and :meth:`DataFrame.to_latex` that would lead to incorrect output when the ``header`` keyword is used (:issue:`16718`)
+- Bug in :meth:`meth:DataFrame.to_csv` where values were truncated when the length of ``na_pep`` was shorter than text input data. (:issue:`25099`)
```
Small typo na_pep -> na_rep

kinow replied:
Ops, fixing it. Thanks!

WillAyd commented on pandas/_libs/writers.pyx:
```diff
@@ -127,19 +127,28 @@ def max_len_string_array(pandas_string[:] arr) -> Py_ssize_t:
+        l = PyString_GET_SIZE(val)
+    elif isinstance(val, bytes):
+        l = PyBytes_GET_SIZE(val)
+    elif isinstance(val, unicode):
```
Don't need this check any more as we dropped Py2 support

kinow replied:
Done! Thanks!

WillAyd commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -6,6 +6,8 @@
 import numpy as np
 import pytest
 
+from pandas.compat import StringIO
```
Can just import from io now

kinow replied:
Good to know it's Py3+ now. Ignoring as this dependency can actually be removed, thanks to your other comment pointing we don't need the `StringIO` at all :+1:

WillAyd commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -561,3 +563,21 @@ def test_to_csv_compression(self, compression_only,
                                  compression=read_compression)
             tm.assert_frame_equal(result, df)
+
+    @pytest.mark.parametrize("df_value", [float('nan')])
```
If we are only passing single values here do we even need parametrization?

kinow replied:
I think the idea was to add more values later, but maybe I never finished the whole work so we didn't get to the point of discussing which new parameters to use?

See https://github.com/pandas-dev/pandas/pull/25103#discussion_r254126828

WillAyd replied:
From that comment I think you just need one parametrization here for dtype, so something like this:

```python
@pytest.mark.parametrize("dtype", [float, 'Int64'])
```

And then do your data frame construction as follows:

```python
df = pd.DataFrame([np.nan] * 3, columns=['c'], dtype=dtype)
```

No need for the astype call thereafter

kinow replied:
Makes sense! Change applied, code rebased. Thanks!

WillAyd commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -561,3 +563,21 @@ def test_to_csv_compression(self, compression_only,
+        expected = tm.convert_rows_list_to_csv_str(expected_rows)
+
+        csv_data = StringIO()
+        df.to_csv(csv_data, index=False, na_rep='mynull', encoding='ascii')
```
Actually you don't even need io - just don't pass an argument for filepath_or_buffer and you should get a string back

kinow replied:
Brilliant!!!

WillAyd reviewed (changes requested):
Can you merge master?

kinow commented:
Thanks for the review @WillAyd 

>Can you merge master?

Done. Rebased, fixed conflicts, and tried to address the feedback, except the parameterized test as that was requested before (I think the intention was to add more test scenarios).

Cheers
Bruno

WillAyd commented on pandas/_libs/writers.pyx:
```diff
@@ -127,19 +127,26 @@ def max_len_string_array(pandas_string[:] arr) -> Py_ssize_t:
+    cdef:
+        Py_ssize_t l = 0
+
+    if isinstance(val, str):
```
Do we even need this any more? Not an expert with the C API but from what I'm reading PyBytes_* is just an alias for PyString_* anyway:

https://docs.python.org/2/c-api/string.html

kinow replied:
Good point. Spent some 10 minutes looking at the Python 3 docs, but not very experienced with the C API, so can't really confirm if that's necessary.

The code in this `if/else` statement is present in [master in another function](https://github.com/pandas-dev/pandas/blob/e02ec8f89a743cc56a79e660b7be850448d15f61/pandas/_libs/writers.pyx#L128-L136). In this pull request I moved it to this other function to re-use it.

WillAyd replied:
I think this should actually be PyUnicode_GET_LENGTH ; PyString is part of the old Python2 interface and there may be some compat provided by Cython best I think safer to use latest and greatest API:

https://docs.python.org/3/howto/cporting.html#str-unicode-unification

https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_GET_LENGTH

kinow replied:
@WillAyd build is passing now :tada:  first I fixed the code without moving to `PyUnicode_GET_LENGTH`, and CI approved it. Then I changed in [this last commit](https://github.com/pandas-dev/pandas/pull/25103/commits/75246bebe144e8ec5cae3a3593062dc727728bb2) to use `PyUnicode_GET_LENGTH`. I never used it before, so hope it was the right way to add it.

- moved what's new to the new file, removing changes from the previous one (we had two releases since this PR was submitted, great work!)
- fixed formatting issues
- used `PyUnicode_GET_LENGTH`

Should be ready for review again.

jreback commented:
can you merge master

simonjayhawkins commented:
@kinow can you merge master.

Conflicting files
doc/source/whatsnew/v0.25.0.rst
pandas/core/internals/blocks.py

kinow commented:
Done @simonjayhawkins :+1:

jreback commented:
this looked reasonble, can you merge master

kinow commented:
Done @jreback 👍

WillAyd reviewed (commented):
Looks like this went for a bit without review - @kinow any chance this is still active?

kinow commented:
Hi @WillAyd 

>Looks like this went for a bit without review - @kinow any chance this is still active?

Sure thing, will just need to re-read last comments and rebase the code. Thanks for the review, I will try to update it today or over the weekend 👍

kinow commented:
Rebased, now will try to find some time this weekend to troubleshoot the Travis errors, and see if I can fix them and them look at `PyUnicode_GET_LENGTH` too.

WillAyd commented on pandas/_libs/writers.pyx:
```diff
@@ -17,6 +12,10 @@ ctypedef fused pandas_string:
 
 
+cdef extern from 'Python.h':
+    Py_ssize_t PyUnicode_GET_LENGTH(object o) nogil
```
Hmm didn't realize this wasn't in Cython. I guess for now stick with `PyUnicode_GET_SIZE` which is there though will be removed in Python 4.0 in favor of this - can deal with that migration comprehensively in our code base later

kinow replied:
> can deal with that migration comprehensively in our code base later

Good idea +1, probably will make it easier to review this PR. Done!

WillAyd commented on pandas/_libs/writers.pyx:
```diff
@@ -125,19 +124,26 @@ def max_len_string_array(pandas_string[:] arr) -> Py_ssize_t:
     return m
 
 
+cpdef inline Py_ssize_t max_len_string(object val):
```
Can you rename this function? It deals with more than just `strings` and returns the length, not the "maximum length". I think just `word_len` or something similar is fine

kinow replied:
Done, now `word_len`, also update the docstring text.

kinow commented:
@WillAyd rebased, fixed conflicts, tried to address your feedback the best I could, and squashed commits too. CI passing, so should be ready for review again 👍 thanks

WillAyd commented on doc/source/whatsnew/v0.25.2.rst:
```diff
@@ -64,7 +64,7 @@ I/O
 - Fix regression in notebook display where <th> tags not used for :attr:`DataFrame.index` (:issue:`28204`).
 - Regression in :meth:`~DataFrame.to_csv` where writing a :class:`Series` or :class:`DataFrame` indexed by an :class:`IntervalIndex` would incorrectly raise a ``TypeError`` (:issue:`28210`)
--
+- Bug in :meth:`DataFrame.to_csv` where values were truncated when the length of ``na_rep`` was shorter than the text input data. (:issue:`25099`)
```
Can you move this to v1.0.0 at this point?

kinow replied:
Sure, done. Moved to v1.0.0 under IO too. Removed the changes to v0.25.2, and edited last commit. Thanks @WillAyd

WillAyd reviewed (changes requested):
Just need to move the whatsnew but I think this is reasonable. Back to @jreback

WillAyd reviewed (approved):
lgtm. By the way @kinow no need to squash your commits we do so before merge

kinow commented:
>lgtm. By the way @kinow no need to squash your commits we do so before merge

Oh, thanks for the heads-up, I thought I would be asked to squash them later. And I realized I had made a mess in the last commits, so decided to tidy it up a bit. Lesson learned for the next PR. Thanks!

TomAugspurger commented:
Looks like we have a couple +1s. Thanks @kinow!
